src/Features/Core/Portable/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,193 @@
+            : base(
+                IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)))
+        {
+        }
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract bool IsWrittenTo(TIdentifierNameSyntax node, SemanticModel model, CancellationToken cancellationToken);
+        protected abstract bool IsMemberOfThisInstance(SyntaxNode node);
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
+
+        protected void AnalyzeType(SyntaxNodeAnalysisContext context)
+        {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CodeStyleOptions.PreferReadonly, context.Node.Language);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var typeSymbol = (ITypeSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node);
+
+            var nonReadonlyFieldMembers = PooledHashSet<IFieldSymbol>.GetInstance();
+            foreach (var item in typeSymbol.GetMembers())
+            {
+                if (item is IFieldSymbol symbol &&
+                    symbol.DeclaredAccessibility == Accessibility.Private &&
+                    !symbol.IsReadOnly &&
+                    !symbol.IsConst &&
+                    !symbol.IsImplicitlyDeclared &&
+                    !IsMutableValueType(symbol.Type))
+                {
+                    var isInCurrentNode = false;
+                    foreach (var syntaxReference in item.DeclaringSyntaxReferences)
+                    {
+                        if (!syntaxReference.SyntaxTree.Equals(context.Node.SyntaxTree))
+                        {
+                            continue;
+                        }
+
+                        if (!context.Node.Span.Contains(syntaxReference.Span))
+                        {
+                            continue;
+                        }
+
+                        isInCurrentNode = true;
+                        break;
+                    }
+
+                    if (isInCurrentNode)
+                    {
+                        nonReadonlyFieldMembers.Add(symbol);
+                    }
+                }
+            }
+
+            var syntaxFactsService = GetSyntaxFactsService();
+            foreach (var syntaxReference in typeSymbol.DeclaringSyntaxReferences)
```

Comment:
>foreach (var syntaxReference in typeSymbol.DeclaringSyntaxReferences) [](start = 12, length = 69)

This seems super convoluted and an extremely expensive n*n algorithm with tons of unnecessary allocations. For a type with `n` partial declarations and `m` field symbols, we will execute this loop `n` times (once for each declaration), allocate this syntax references array `n` times (potentially for every named type in the compilation). We also allocate n*n semantic model instances (compilation does not cache semantic model for trees), and invoke very similar semantic queries on different instances of these semantic models (each adds to binding cost).

An IOperation based implementation will analyze every field reference and type declaration exactly once, without any of these performance overheads. I think it is worth spending half a day converting this to IOperation analyzer then shipping an implementation with known performance problems.

Reply:
>  allocate this syntax references array n times

Can you clarify that bit.  It seems like this is only being allocated once per type.

Reply:
Actually, i think i see what you're saying.  because thsi is a syntax-node algorithm, we'll both hit all the declaration nodes for a partial type.  And then for each part, we'll then go examine all the other parts.  Yes, that seems wonky.  It seems like this could either be made into an IOperation analyzer (which likely wouldn't be that hard), or it could be a symbol analyzer that registers for INamedTypeSymbols.

Reply:
> Can you clarify that bit.  It seems like this is only being allocated once per type.  

The analyzer is registering a syntax node action, not a symbol action. So there will be `n` callbacks, once per each partial declaration of the containing type and each callback can potentially execute this loop.

---
In reply to: [176219860](https://github.com/dotnet/roslyn/pull/19067#discussion_r176219860) [](ancestors = 176219860)

Reply:
Yup.  Got it.  That does seem wonky.